<a href="https://colab.research.google.com/github/thakursejal/LLM-APIs-AI-Chat-Application/blob/main/LLM_APIs_AI_Chat_Application_Thakur_Sejal.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# MODULE 3 — SUBPOINT 1
# LLM APIs & AUTHENTICATION

**Name:** Thakur Sejal

---

# 1. Introduction to LLM APIs

A Large Language Model (LLM) API allows applications to communicate with an AI language model programmatically.

Instead of interacting with an AI model only through a website, developers can send requests from their own applications and receive AI-generated responses.

LLM APIs can be used to build:

- AI chatbots
- Virtual assistants
- Content generation tools
- Summarization applications
- Question-answering systems
- Coding assistants
- Educational applications

---

# 2. How an LLM API Works

A basic API workflow is:

**Application → API Request → LLM → API Response → Application**

The application sends information to the API.

The API sends the request to the selected language model.

The model processes the request and generates a response.

The response is then returned to the application.

---

# 3. Important API Concepts

## API Key

An API key is a credential used to authenticate an application with an API service.

API keys should be treated as sensitive information.

Never publish an API key in:

- GitHub repositories
- Screenshots
- README files
- Public notebooks
- Social media posts

---

## Authentication

Authentication verifies that an application is authorized to access an API.

A common approach is:

**Application → API Key → API Service → Authentication → Response**

---

## Environment Variables

Environment variables provide a safer way to store sensitive configuration values such as API keys.

Instead of writing a key directly inside source code, applications can retrieve it from an environment variable.

Example:

`API_KEY`

The actual secret value should remain private.

---

# 4. Why API Keys Should Be Protected

Exposing an API key can allow unauthorized users to use the associated API account.

Possible consequences include:

- Unauthorized API usage
- Unexpected charges
- Account security problems
- Abuse of API resources

Therefore, API keys should never be hard-coded into publicly shared projects.

---

# 5. LLM API Request Structure

A typical LLM request may contain:

- Model name
- User message
- System instructions
- Conversation history
- Generation parameters

A simplified workflow is:

**Input → API Request → Model Processing → Generated Response**

---

# 6. System, User and Assistant Messages

Many chat-based LLM APIs use different message roles.

## System Message

Defines the behavior or instructions for the AI.

Example:

"You are a helpful programming tutor."

## User Message

Contains the user's request.

Example:

"Explain Python functions."

## Assistant Message

Contains the AI-generated response.

Example:

"Python functions are reusable blocks of code..."

---

# 7. Model Parameters

LLM APIs may provide parameters that influence generation.

Important parameters can include:

### Temperature

Controls the randomness of generated responses.

Lower values generally produce more predictable responses.

Higher values can produce more varied responses.

### Maximum Output Tokens

Controls the maximum amount of generated output.

### Model

Specifies which language model should process the request.

The available parameters depend on the API provider and model.

---

# 8. Basic LLM API Workflow in Python

A Python application generally follows these steps:

1. Install or import the required SDK.
2. Obtain an API key securely.
3. Configure authentication.
4. Select an LLM model.
5. Create the request.
6. Send the request.
7. Receive the response.
8. Display or process the response.
9. Handle possible errors.

---

# 9. Secure Development Practice

A safe development workflow is:

**Secret → Environment Variable → Python Application → API Request**

The secret should not be included directly in the source code.

For public projects, sensitive credentials should also be excluded from Git commits.

---

# 10. Key Learning

LLM APIs allow developers to integrate language models into their own applications.

Authentication is required to securely access an API.

Environment variables provide a safer way to manage API credentials.

Understanding API requests, responses, message roles and model parameters is essential for building reliable LLM applications.

---

# END OF SUBPOINT 1

In [ ]:
!pip install -q -U google-genai

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56.3/56.3 kB 2.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.1/1.1 MB 21.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 262.4/262.4 kB 22.6 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
google-colab 1.0.0 requires google-auth==2.49.0, but you have google-auth 2.58.0 which is incompatible.


In [ ]:
from google import genai

print("Gemini SDK installed successfully!")

Gemini SDK installed successfully!


In [5]:
import os
from getpass import getpass

hf_token = getpass("Enter your Hugging Face token: ")
os.environ["HF_TOKEN"] = hf_token

print("Hugging Face token loaded securely for this session.")

Enter your Hugging Face token: ··········
Hugging Face token loaded securely for this session.


In [6]:
!pip install -q -U huggingface_hub

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 842.9/842.9 kB 12.8 MB/s eta 0:00:00


In [10]:
from huggingface_hub import InferenceClient
import os

client = InferenceClient(
    api_key=os.environ["HF_TOKEN"]
)

print("Hugging Face API client initialized successfully!")

Hugging Face API client initialized successfully!


In [11]:
response = client.chat.completions.create(
    model="openai/gpt-oss-120b:fastest",
    messages=[
        {
            "role": "user",
            "content": "Explain artificial intelligence in simple words for a beginner."
        }
    ],
    max_tokens=150
)

print(response.choices[0].message.content)

**Artificial Intelligence (AI) – A Beginner’s Friendly Guide**

---

### 1. What is AI?  
Think of AI as a *computer program that can do tasks that usually need a human brain*.  
- **“Artificial”** → made by people, not natural.  
- **“Intelligence”** → the ability to learn, reason, understand language, see


# 11. Practical LLM API Experiment

## Objective

The objective of this experiment was to connect a Python application to a Large Language Model using an API and obtain an AI-generated response.

## API Workflow

**Python Application → Authentication → Hugging Face API → Inference Provider → LLM → AI Response**

## Steps Performed

1. Installed the Hugging Face Python client.
2. Created a Hugging Face access token with inference permissions.
3. Loaded the token securely using an environment variable.
4. Initialized the Hugging Face InferenceClient.
5. Selected a supported chat-completion model.
6. Sent a user prompt through the API.
7. Received and displayed the generated AI response.

## Example Prompt

"Explain artificial intelligence in simple words for a beginner."

## Result

The API successfully processed the request and returned an AI-generated response.

## Observation

The experiment demonstrated how an external LLM can be integrated into a Python application using an API.

The API token provides authentication, while the selected model processes the user message and returns the generated response.

## Security Practice

The Hugging Face token was entered securely during the Colab session.

The token was not included directly in the source code and should not be uploaded to GitHub.

## Key Learning

This experiment provided practical experience with:

- LLM APIs
- API authentication
- Environment variables
- Python API clients
- Chat-completion requests
- Model selection
- AI-generated responses
- Secure credential handling

---

# SUBPOINT 1 COMPLETED

**LLM APIs & Authentication**

The basic LLM API connection has been successfully demonstrated using Python and Hugging Face.

# MODULE 3 — SUBPOINT 2
# CONVERSATION MANAGEMENT & CHAT HISTORY

**Name:** Thakur Sejal

---

# 12. Introduction to Conversation Management

Conversation management is the process of maintaining and organizing messages exchanged between a user and an AI model.

A conversational AI application usually needs to remember previous messages so that the model can understand the context of the conversation.

A typical conversation contains:

- System messages
- User messages
- Assistant messages

The conversation can be represented as a sequence of messages.

---

# 13. Message Roles

## System Message

A system message defines the behavior, role or instructions for the AI.

Example:

"You are a helpful AI tutor."

## User Message

A user message contains the request or question from the user.

Example:

"Explain machine learning."

## Assistant Message

An assistant message contains the response generated by the AI.

Example:

"Machine learning is a method that allows computers to learn patterns from data."

---

# 14. Chat History

Chat history stores previous messages exchanged during a conversation.

Example:

User:
"Explain Python."

Assistant:
"Python is a programming language..."

User:
"What is a variable?"

The previous conversation helps the AI understand that the user is still asking about Python.

---

# 15. Why Chat History Is Important

Chat history helps applications provide:

- Context-aware responses
- Follow-up question support
- Continuous conversations
- Personalized interactions
- Better understanding of previous messages

Without conversation history, each request may be treated as an independent question.

---

# 16. Managing Conversation History in Python

A simple conversation history can be stored using a Python list.

Example structure:

    messages = [
        {
            "role": "system",
            "content": "You are a helpful AI tutor."
        },
        {
            "role": "user",
            "content": "What is Python?"
        },
        {
            "role": "assistant",
            "content": "Python is a programming language."
        }
    ]

New messages can be added to the list as the conversation continues.

---

# 17. Conversation Flow

A basic conversational workflow is:

**User Input → Add to History → Send History to LLM → Receive Response → Add Response to History**

This process is repeated for every new user message.

---

# 18. System, User and Assistant Example

A conversation can look like:

System:
"You are an AI study assistant."

User:
"Explain neural networks."

Assistant:
"Neural networks are computational models inspired by the human brain."

User:
"Give me a simple example."

Assistant:
"An image classification system is a common example."

The assistant can use the previous messages to understand the context.

---

# 19. Model Parameters

LLM APIs provide parameters that control how responses are generated.

Common parameters include:

## Model

Specifies which language model should process the request.

## Maximum Output Tokens

Controls the maximum amount of text generated by the model.

## Temperature

Controls randomness in model responses when supported by the selected API/model.

Lower values generally produce more predictable responses.

Higher values can produce more varied responses.

The exact parameters available depend on the API provider and model.

---

# 20. Token Usage

Tokens are pieces of text processed by an LLM.

For example, a word may be represented as one or more tokens depending on the tokenizer.

LLM APIs may measure:

- Input tokens
- Output tokens
- Total tokens

Token usage can be important for:

- API cost management
- Context-window management
- Performance optimization
- Limiting unnecessary conversation history

---

# 21. Context Window

A context window is the amount of information a model can consider in a single request.

Long conversations can consume more tokens.

Therefore, applications may need to:

- Limit conversation history
- Summarize older messages
- Remove unnecessary messages
- Control maximum output length

---

# 22. Conversation Management Best Practices

Good conversation management includes:

1. Store messages in a structured format.
2. Preserve important context.
3. Avoid unnecessary messages.
4. Limit excessive history.
5. Handle errors gracefully.
6. Monitor token usage.
7. Protect sensitive information.
8. Provide clear system instructions.

---

# 23. Key Learning

Conversation management allows an AI application to maintain context across multiple user interactions.

Chat history can be stored as structured messages containing system, user and assistant roles.

Model parameters influence how the AI generates responses.

Token usage and context limits should be considered when building real-world conversational applications.

---

# END OF SUBPOINT 2 THEORY

In [12]:
# Conversation Management Experiment

messages = [
    {
        "role": "system",
        "content": "You are a helpful AI study assistant. Explain concepts clearly for a B.Tech student."
    },
    {
        "role": "user",
        "content": "What is machine learning?"
    }
]

response = client.chat.completions.create(
    model="openai/gpt-oss-120b:fastest",
    messages=messages,
    max_tokens=150
)

assistant_reply = response.choices[0].message.content

print("Assistant:", assistant_reply)

# Add the assistant response to chat history
messages.append({
    "role": "assistant",
    "content": assistant_reply
})

print("\n--- Chat History ---")
for message in messages:
    print(message["role"].upper() + ":", message["content"])

Assistant: **Machine Learning (ML) – A B‑Tech‑Friendly Overview**

---

### 1. What is Machine Learning?
Machine Learning is a **sub‑field of Artificial Intelligence (AI)** that focuses on building systems that **learn from data** rather than being explicitly programmed with rules.  
In other words, instead of writing a line‑by‑line algorithm for every possible situation, we give the

--- Chat History ---
SYSTEM: You are a helpful AI study assistant. Explain concepts clearly for a B.Tech student.
USER: What is machine learning?
ASSISTANT: **Machine Learning (ML) – A B‑Tech‑Friendly Overview**

---

### 1. What is Machine Learning?
Machine Learning is a **sub‑field of Artificial Intelligence (AI)** that focuses on building systems that **learn from data** rather than being explicitly programmed with rules.  
In other words, instead of writing a line‑by‑line algorithm for every possible situation, we give the


In [13]:
# Follow-up Question Using Chat History

messages.append({
    "role": "user",
    "content": "Can you give me a simple real-world example?"
})

response = client.chat.completions.create(
    model="openai/gpt-oss-120b:fastest",
    messages=messages,
    max_tokens=150
)

assistant_reply = response.choices[0].message.content

print("Assistant:", assistant_reply)

# Add the new assistant response to the conversation history
messages.append({
    "role": "assistant",
    "content": assistant_reply
})

print("\n--- Updated Chat History ---")
for message in messages:
    print(message["role"].upper() + ":", message["content"])

Assistant: ### A Simple Real‑World Example – **Spam‑Email Filter**

One of the classic, easy‑to‑understand applications of machine learning is **automatically deciding whether an incoming email is “spam” or “not‑spam.”**  
Below we walk through the whole ML pipeline using a very small dataset, so you can see every step in action.

---

## 1. The Problem (What we want to solve)

| Input (

--- Updated Chat History ---
SYSTEM: You are a helpful AI study assistant. Explain concepts clearly for a B.Tech student.
USER: What is machine learning?
ASSISTANT: **Machine Learning (ML) – A B‑Tech‑Friendly Overview**

---

### 1. What is Machine Learning?
Machine Learning is a **sub‑field of Artificial Intelligence (AI)** that focuses on building systems that **learn from data** rather than being explicitly programmed with rules.  
In other words, instead of writing a line‑by‑line algorithm for every possible situation, we give the
USER: Can you give me a simple real-world example?
ASSISTANT: ### A

In [14]:
# Model Parameters and Token Usage

response = client.chat.completions.create(
    model="openai/gpt-oss-120b:fastest",
    messages=[
        {
            "role": "system",
            "content": "You are a concise AI tutor."
        },
        {
            "role": "user",
            "content": "Explain neural networks in simple words."
        }
    ],
    max_tokens=100
)

print("AI Response:")
print(response.choices[0].message.content)

print("\n--- Usage Information ---")

if response.usage:
    print("Input tokens:", response.usage.prompt_tokens)
    print("Output tokens:", response.usage.completion_tokens)
    print("Total tokens:", response.usage.total_tokens)
else:
    print("Token usage information was not returned by the provider.")

AI Response:
**Neural networks** are a type of computer program that learns to recognize patterns—much like a brain does.  

---

### 1. The building blocks: **Artificial

--- Usage Information ---
Input tokens: 88
Output tokens: 100
Total tokens: 188


# 24. Conversation Management Practical Experiment

## Objective

The objective of this experiment was to understand how an LLM application can maintain conversation context using chat history.

## Experiment 1 — Initial Conversation

A system message was created to define the AI as a study assistant.

A user question was then sent to the model:

"What is machine learning?"

The generated assistant response was stored in the conversation history.

---

## Experiment 2 — Follow-Up Question

A second user message was added:

"Can you give me a simple real-world example?"

The previous conversation history was sent together with the new question.

The model generated a context-aware response.

### Conversation Flow

**System Instruction → User Question → Assistant Response → User Follow-Up → Context-Aware Assistant Response**

---

## Experiment 3 — Token Usage

A separate API request was used to observe token usage.

The response provided information about:

- Input tokens
- Output tokens
- Total tokens

The maximum output length was also controlled using the `max_tokens` parameter.

---

## Observations

The experiments demonstrated that:

- Chat history allows the model to maintain conversation context.
- System messages can define the behavior of the AI.
- User messages represent requests.
- Assistant messages store generated responses.
- Previous messages can be sent with a new request to support follow-up questions.
- Token usage can be monitored when supported by the API provider.
- Limiting output tokens can help control response length.

---

## Conversation History Structure

The conversation was maintained using a Python list containing dictionaries.

Each message included:

- `role`
- `content`

Example roles:

- `system`
- `user`
- `assistant`

This structure can be used as the foundation for a conversational AI application.

---

## Key Learning

Conversation management is an essential part of building an AI chatbot.

Maintaining structured chat history allows the application to provide context-aware responses.

Model parameters can control aspects of the generated response, while token usage provides useful information for managing context and API resources.

---

# SUBPOINT 2 COMPLETED

**Conversation Management, Chat History & Model Parameters**

# MODULE 3 — SUBPOINT 3
# STRUCTURED RESPONSES, ERROR HANDLING & API COST MANAGEMENT

**Name:** Thakur Sejal

---

# 25. Introduction to Structured Responses

A structured response is an AI response organized in a predefined format.

Instead of receiving only plain text, an application can request information in a predictable structure.

Common structured formats include:

- JSON
- Lists
- Dictionaries
- Tables
- Key-value pairs

Structured responses are useful because they make AI-generated information easier for software applications to process.

---

# 26. Why Structured Responses Are Useful

Structured responses can help with:

- Data processing
- Application integration
- Information extraction
- Automation
- Database storage
- API communication
- Consistent application outputs

For example, instead of receiving:

"Sejal is a B.Tech student studying Artificial Intelligence and Machine Learning."

An application could organize the information as:

Name: Sejal  
Course: B.Tech  
Branch: Artificial Intelligence and Machine Learning

---

# 27. JSON Responses

JSON stands for JavaScript Object Notation.

It is a lightweight format commonly used for exchanging structured data.

A JSON object generally contains key-value pairs.

Example:

{
    "name": "Sejal",
    "course": "B.Tech",
    "branch": "AIML"
}

Applications can easily read and process this information.

---

# 28. Structured Output Workflow

A basic structured-output workflow is:

**User Input → LLM → Structured Response → Application Processing**

The application can then extract individual fields from the response.

---

# 29. Error Handling

API requests can fail for different reasons.

Possible causes include:

- Invalid authentication
- Network problems
- Unsupported models
- Invalid request parameters
- Rate limits
- Service availability problems
- Incorrect input
- Provider restrictions

Applications should handle these situations gracefully instead of crashing.

---

# 30. Basic Error Handling Strategy

A simple Python approach is:

**Try → API Request → Success**

or:

**Try → API Request → Error → Handle Exception → User-Friendly Message**

The user should receive a clear message instead of a confusing technical traceback whenever possible.

---

# 31. Why Error Handling Matters

Good error handling helps applications become:

- More reliable
- Easier to debug
- More user-friendly
- More maintainable
- More resilient

A production AI application should anticipate API failures and handle them appropriately.

---

# 32. Basic API Cost Management

LLM API usage can consume resources and may have usage limits or costs depending on the provider and model.

Important factors include:

- Number of API requests
- Input tokens
- Output tokens
- Model selection
- Conversation history
- Maximum output length

---

# 33. Techniques for Managing API Usage

## 1. Limit Output Length

Use an appropriate maximum output token limit.

## 2. Control Conversation History

Avoid sending unnecessary old messages.

## 3. Use Appropriate Models

Choose a model suitable for the task instead of unnecessarily using a larger model.

## 4. Avoid Repeated Requests

Do not send duplicate API requests unnecessarily.

## 5. Monitor Token Usage

Track input and output token usage when available.

## 6. Validate User Input

Avoid sending empty or meaningless requests.

---

# 34. Reliable AI Application Workflow

A reliable AI application can follow:

**User Input → Validation → API Request → Error Handling → Response Validation → Display Result**

This workflow reduces unexpected failures.

---

# 35. Security Considerations

API credentials should be protected.

Important practices include:

- Never hard-code API tokens in public source code.
- Never upload tokens to GitHub.
- Use environment variables or secret management.
- Avoid sharing credentials in screenshots.
- Revoke compromised tokens.
- Use minimum required permissions.

---

# 36. Key Learning

Structured responses make AI outputs easier for applications to process.

Error handling helps applications respond gracefully when API requests fail.

Basic API cost management involves controlling requests, tokens, conversation history and model usage.

Secure credential management is essential when developing applications that communicate with external AI services.

---

# END OF SUBPOINT 3 THEORY

In [15]:
# Structured JSON Response Experiment

prompt = """
Return information about Artificial Intelligence in JSON format.

Use exactly these fields:
{
  "definition": "short definition",
  "applications": ["application 1", "application 2", "application 3"],
  "benefits": ["benefit 1", "benefit 2"],
  "beginner_tip": "one simple learning tip"
}

Return only valid JSON without explanations or Markdown.
"""

response = client.chat.completions.create(
    model="openai/gpt-oss-120b:fastest",
    messages=[
        {
            "role": "user",
            "content": prompt
        }
    ],
    max_tokens=250
)

print("Structured AI Response:")
print(response.choices[0].message.content)

Structured AI Response:
None


In [16]:
# Inspect the API response structure safely

print("Response object:")
print(response)

print("\nChoices:")
print(response.choices)

print("\nFirst message:")
print(response.choices[0].message)

Response object:
ChatCompletionOutput(choices=[ChatCompletionOutputComplete(finish_reason='length', index=0, message=ChatCompletionOutputMessage(role='assistant', content=None, reasoning='The user wants information about Artificial Intelligence in JSON format with specific fields. Must return only valid JSON without explanations or markdown.\n\nFields:\n- "definition": short definition\n- "applications": array with three items (application 1,2,3) presumably about AI.\n- "benefits": array with two items\n- "beginner_tip": string\n\nNeed to provide actual content, not placeholders. So we must fill with actual data.\n\nDefinition: short definition of AI. Applications: three common AI applications (e.g., image recognition, natural language processing, recommendation systems). Benefits: two benefits (e.g., increased efficiency, improved decision-making). Beginner tip: one simple learning tip (e.g., start with Python and basic machine learning tutorials).\n\nMake sure JSON is valid: keys in 

In [17]:
# Corrected Structured JSON Response Experiment

prompt = """
Return information about Artificial Intelligence as valid JSON.

Use exactly these fields:
definition
applications
benefits
beginner_tip

Keep each value short.
Return ONLY the JSON object.
"""

response = client.chat.completions.create(
    model="openai/gpt-oss-120b:fastest",
    messages=[
        {
            "role": "user",
            "content": prompt
        }
    ],
    max_tokens=500
)

content = response.choices[0].message.content

print("Structured AI Response:")
print(content)

Structured AI Response:
{
  "definition": "AI is the simulation of human intelligence by machines, enabling them to learn, reason, and act.",
  "applications": "Automation, healthcare diagnostics, finance, language translation, and recommendation systems.",
  "benefits": "Increases efficiency, uncovers insights, reduces errors, and enables new services.",
  "beginner_tip": "Start with simple projects using libraries like scikit-learn or TensorFlow to grasp core concepts."
}


In [18]:
# API Error Handling Experiment

try:
    response = client.chat.completions.create(
        model="invalid/model-name",
        messages=[
            {
                "role": "user",
                "content": "Explain machine learning in simple words."
            }
        ],
        max_tokens=100
    )

    print("AI Response:")
    print(response.choices[0].message.content)

except Exception as error:
    print("API request failed.")
    print("Error handled successfully.")
    print("Error type:", type(error).__name__)

API request failed.
Error handled successfully.
Error type: BadRequestError


In [19]:
# API Usage & Token Management Experiment

response = client.chat.completions.create(
    model="openai/gpt-oss-120b:fastest",
    messages=[
        {
            "role": "system",
            "content": "You are a concise AI tutor."
        },
        {
            "role": "user",
            "content": "Explain three applications of artificial intelligence."
        }
    ],
    max_tokens=120
)

print("AI Response:")
print(response.choices[0].message.content)

print("\n--- Token Usage ---")

if response.usage:
    print("Input tokens:", response.usage.prompt_tokens)
    print("Output tokens:", response.usage.completion_tokens)
    print("Total tokens:", response.usage.total_tokens)
else:
    print("Token usage information was not returned by the provider.")

AI Response:
**1. Healthcare – AI‑assisted diagnostics**  
Machine‑learning models (e.g., deep‑learning image classifiers) can examine X‑rays, MRIs, or pathology slides to flag abnormalities such as tumors or fractures. By ranking the likelihood

--- Token Usage ---
Input tokens: 88
Output tokens: 120
Total tokens: 208


# 37. Subpoint 3 Practical Experiments

## Experiment 1 — Structured JSON Response

A structured JSON response was generated by providing the AI with a predefined output format.

### Objective

To demonstrate how an LLM can return information in a format that can be processed by an application.

### Result

The model successfully generated a structured JSON response containing predefined fields.

### Observation

Structured responses make AI-generated information easier to organize, process and integrate into software applications.

---

## Experiment 2 — API Error Handling

An invalid model name was intentionally used to simulate an API failure.

Python's `try-except` mechanism was used to catch the error.

### Result

The API error was detected and handled without allowing the application to terminate unexpectedly.

### Observation

Error handling improves application reliability and provides users with clearer feedback when API requests fail.

---

## Experiment 3 — Token Usage Management

An LLM request was made with a maximum output limit.

The response usage information was examined.

### Observed Token Usage

- Input tokens: 88
- Output tokens: 120
- Total tokens: 208

### Observation

Monitoring token usage can help developers understand API consumption and manage application resources.

Limiting output length can also help prevent unnecessarily long responses.

---

# 38. Practical Error Handling Pattern

A reliable LLM application can follow:

**Validate Input → Send API Request → Handle Error → Validate Response → Display Result**

This approach helps create more robust applications.

---

# 39. API Usage Management

Important techniques include:

- Limiting output length
- Monitoring token usage
- Avoiding unnecessary requests
- Managing conversation history
- Selecting an appropriate model
- Validating user input
- Handling API errors

---

# 40. Subpoint 3 Key Learning

This practical work demonstrated three important aspects of LLM application development:

### Structured Responses

AI output can be organized into predictable formats such as JSON.

### Error Handling

API failures should be caught and handled gracefully.

### API Usage Management

Token usage and response length can be monitored to improve resource management.

Secure handling of API credentials is also essential when working with external AI services.

---

# SUBPOINT 3 COMPLETED

**Structured Responses, Error Handling & Basic API Cost Management**

# MODULE 3 — SUBPOINT 4
# BUILD AN AI CHAT APPLICATION

**Name:** Thakur Sejal

---

# 41. Introduction to AI Chat Applications

An AI chat application is a software application that allows users to communicate with a Large Language Model through a conversational interface.

The application accepts user input, sends it to an LLM through an API, receives the generated response and displays it to the user.

A basic AI chatbot follows this workflow:

**User Input → Chat History → LLM API → AI Response → Updated History → Display**

---

# 42. Main Components

The AI chat application developed in this module contains:

- User input
- System instructions
- LLM API connection
- Chat history
- AI-generated responses
- Error handling
- Basic user interface

---

# 43. User Input

The application should allow the user to enter a question or instruction.

Example:

"Explain neural networks in simple words."

The input is added to the conversation history before being sent to the LLM.

---

# 44. Custom Instructions

Custom instructions allow the user or developer to define how the AI should behave.

Example:

"You are a helpful AI tutor. Explain technical concepts clearly and use simple examples."

Custom instructions can control:

- Role
- Tone
- Explanation style
- Target audience
- Response behavior

---

# 45. Chat History

The application maintains a list of messages.

Each message contains:

- Role
- Content

The main roles are:

- System
- User
- Assistant

Maintaining chat history allows the AI to understand previous messages and answer follow-up questions with context.

---

# 46. LLM API Integration

The application sends the conversation history to the LLM API.

The API processes the request and returns an AI-generated response.

The response is then added to the chat history.

---

# 47. Error Handling

The application should handle possible API errors.

Examples include:

- Authentication errors
- Network errors
- Invalid requests
- Unsupported models
- Provider errors
- Rate limits

A user-friendly error message should be displayed instead of an unexplained technical traceback.

---

# 48. Basic User Interface

A simple chat interface can contain:

1. Application title
2. Custom instructions
3. Chat history
4. User input box
5. Send button
6. AI response area
7. Clear conversation option

---

# 49. Streamlit Application

Streamlit can be used to convert the Python chatbot into an interactive web application.

The basic workflow is:

**Python Code → GitHub Repository → Streamlit Community Cloud → Live Web Application**

The deployed application can then be accessed through a public Streamlit URL.

---

# 50. Application Features

The final AI chat application will include:

- 🤖 AI chatbot
- 💬 Multi-turn conversation
- 🧠 Conversation history
- ⚙️ Custom system instructions
- 🔐 Secure API token handling
- 🛡️ Error handling
- 🧹 Clear chat option
- 🌐 Web-based interface

---

# 51. Development Workflow

The application will be developed using the following process:

**Design → API Integration → Chat History → UI → Error Handling → Testing → GitHub → Streamlit Deployment**

---

# 52. Key Learning

Building an AI chat application combines the concepts learned throughout Module 3.

The application uses:

- LLM APIs
- Authentication
- Environment variables
- Message roles
- Chat history
- Model parameters
- Structured responses
- Error handling
- Token management
- User interface development

---

# SUBPOINT 4 — APPLICATION DEVELOPMENT

In [20]:
# AI Chat Application - Colab Prototype

import os
from huggingface_hub import InferenceClient

client = InferenceClient(
    api_key=os.environ["HF_TOKEN"]
)

SYSTEM_INSTRUCTION = """
You are a helpful AI study assistant for a B.Tech AIML student.
Explain technical concepts clearly using simple language and examples.
"""

chat_history = [
    {
        "role": "system",
        "content": SYSTEM_INSTRUCTION
    }
]

def chat_with_ai(user_message):
    chat_history.append({
        "role": "user",
        "content": user_message
    })

    try:
        response = client.chat.completions.create(
            model="openai/gpt-oss-120b:fastest",
            messages=chat_history,
            max_tokens=300
        )

        assistant_reply = response.choices[0].message.content

        if not assistant_reply:
            assistant_reply = "The AI did not return a response."

        chat_history.append({
            "role": "assistant",
            "content": assistant_reply
        })

        return assistant_reply

    except Exception as error:
        return f"Sorry, an error occurred: {type(error).__name__}"


print("AI Chat Application Ready!")
print("Type a question when prompted.")

user_question = input("You: ")

answer = chat_with_ai(user_question)

print("\nAI:", answer)

AI Chat Application Ready!
Type a question when prompted.
You: Explain machine learning in simple words.

AI: **Machine Learning (ML) – A Simple Everyday Description**

---

### 1. What It Is  
Imagine you have a smart helper that **learns from experience** instead of following a fixed list of instructions.  
- **Traditional programming:** You write exact rules (“If the email contains “free”, mark it as spam”).  
- **Machine learning:** You give the helper lots of examples (“These emails are spam, these are not”) and it figures out the hidden patterns on its own.  

In short, **machine learning is a way to let computers improve at a task by learning from data**.

---

### 2. How It Works – The Basic Steps  

| Step | What Happens (Simple View) | Everyday Analogy |
|------|----------------------------|------------------|
| **1️⃣ Collect Data** | Gather examples of the problem (pictures, text, numbers). | Like collecting photos of cats and dogs. |
| **2️⃣ Choose a Model** | Pick a “math 

In [21]:
# Clear Chat History

def clear_chat():
    global chat_history

    chat_history = [
        {
            "role": "system",
            "content": SYSTEM_INSTRUCTION
        }
    ]

    print("Chat history cleared successfully.")


# Test the clear-chat feature
clear_chat()

print("\nCurrent Chat History:")
print(chat_history)

Chat history cleared successfully.

Current Chat History:
[{'role': 'system', 'content': '\nYou are a helpful AI study assistant for a B.Tech AIML student.\nExplain technical concepts clearly using simple language and examples.\n'}]


In [22]:
# Multi-Turn Conversation Test

clear_chat()

questions = [
    "What is machine learning?",
    "Give me one real-world example.",
    "What is the main benefit of that example?"
]

for question in questions:
    print("\nYou:", question)

    answer = chat_with_ai(question)

    print("AI:", answer)

print("\n--- Final Chat History ---")

for message in chat_history:
    print(message["role"].upper() + ":", message["content"])

Chat history cleared successfully.

You: What is machine learning?
AI: **Machine Learning (ML) – A Simple Overview**

| Aspect | What it means in everyday language |
|--------|--------------------------------------|
| **Definition** | *Teaching a computer to improve at a task by learning from data, instead of giving it step‑by‑step instructions.* |
| **Why it matters** | It lets computers solve problems that are hard to describe with exact rules – e.g., recognizing faces, translating languages, or recommending a movie you’ll like. |

---

## 1. How ML Works – The Basic Cycle

1. **Collect Data** – Gather examples of the problem you want to solve.  
   *Example:* Thousands of emails labeled “spam” or “not‑spam”.  

2. **Pick a Model** – Choose a mathematical formula (a “model”) that can map inputs to outputs.  
   *Example:* A logistic‑regression model,

You: Give me one real-world example.
AI: ### Real‑World Example: **Spam‑Detection for Email Services**

---

#### 1. The Problem  
Whe

# 53. AI Chat Application — Practical Testing

## Objective

The objective of this experiment was to build a basic conversational AI application using Python and an LLM API.

The application accepts user input, sends the conversation to the LLM and displays the generated response.

---

## Main Features Tested

### 1. User Input

The application accepts questions from the user through Python input.

### 2. AI Response

The user message is sent to the Hugging Face LLM API and the generated response is displayed.

### 3. Conversation History

User and assistant messages are stored in a structured chat-history list.

### 4. Custom Instructions

A system instruction defines the AI as a helpful study assistant for a B.Tech AIML student.

### 5. Multi-Turn Conversation

Multiple questions can be asked in the same conversation.

Previous messages are included in later API requests so the model can use conversation context.

### 6. Clear Chat

A clear-chat function resets the conversation while preserving the system instruction.

### 7. Error Handling

API exceptions are caught and converted into a simple user-friendly message.

---

# 54. Application Workflow

**User Input → Validate/Process Input → Add to Chat History → LLM API → AI Response → Add Response to History → Display Response**

---

# 55. Multi-Turn Conversation Test

The chatbot was tested with multiple questions:

1. What is machine learning?
2. Give me one real-world example.
3. What is the main benefit of that example?

The application maintained the conversation history while processing the questions.

---

# 56. Testing Results

| Feature | Test Result |
|---|---|
| LLM API Connection | ✅ Successful |
| User Input | ✅ Successful |
| AI Response | ✅ Successful |
| System Instructions | ✅ Successful |
| Chat History | ✅ Successful |
| Multi-Turn Conversation | ✅ Successful |
| Clear Chat | ✅ Successful |
| Error Handling | ✅ Implemented |
| Token Management | ✅ Demonstrated |

---

# 57. Application Development Learning

Building the chatbot combined the concepts learned throughout Module 3:

- LLM API integration
- Authentication
- Environment variables
- Message roles
- Conversation history
- Model parameters
- Structured responses
- Error handling
- Token usage
- Custom instructions
- Basic application development

---

# 58. Streamlit Deployment Plan

The Colab prototype will now be converted into a web application using Streamlit.

The deployment workflow will be:

**Python Application → GitHub Repository → Streamlit Community Cloud → Live AI Chat Application**

The Hugging Face API token will be stored as a deployment secret rather than being included in the public source code.

---

# SUBPOINT 4 COMPLETED

**AI Chat Application Development**

The Colab prototype has been successfully developed and tested.

The next stage is to create the Streamlit web interface and deploy the application.